### PR TITLE
Add userdefaut UserDefault::deleteValueForKey

### PR DESCRIPTION
**Pull request by jianglong0156**
Add userdefaut UserDefault::deleteValueForKey

already test in android, win32, ios, mac, winrt

### CODE REVIEW

**Comment by pandamicro** [comment]
@andyque Please review this PR

@jianglong0156 Please link this PR to related issue

**Comment by jianglong0156** [comment]
sorry for my misstake. the related issue is https://github.com/cocos2d/cocos2d-x/issues/9878

File: cocos/base/CCUserDefault-android.cpp [hunk 1]
```diff
@@ -519,6 +519,17 @@ void UserDefault::flush()
 {
 }
 
+bool UserDefault::deleteValueForKey(const char* key)
+{
+    // check the params
+    if (!key)
```
Comment (zilongshanren) [inline cocos/base/CCUserDefault-android.cpp:525]: It's better to output some debug info to tell why it fails.
Like `CCLOG("key can't be nullptr!")`.

File: cocos/base/CCUserDefault-apple.mm [hunk 2]
```diff
@@ -542,6 +542,20 @@ static void deleteNodeByKey(const char *pKey)
     [[NSUserDefaults standardUserDefaults] synchronize];
 }
 
+bool UserDefault::deleteValueForKey(const char* key)
+{
+    // check the params
+    if (!key)
```
Comment (zilongshanren) [inline cocos/base/CCUserDefault-apple.mm:548]: The same as above comments.

File: cocos/base/CCUserDefault-android.cpp [hunk 3]
```diff
@@ -519,6 +519,17 @@ void UserDefault::flush()
...
+    {
+        return false;
+    }
+    
+    deleteValueForKeyJNI(key);
```
Comment (zilongshanren) [inline cocos/base/CCUserDefault-android.cpp:530]: missing a `flush` function here?

File: cocos/base/CCUserDefault-winrt.cpp [hunk 4]
```diff
@@ -329,6 +329,21 @@ void UserDefault::flush()
...
+{
+    // check the params
+    if (!key)
+    {
+        return false;
+    }
+
+    ApplicationDataContainer^ localSettings = ApplicationData::Current->LocalSettings;
+    auto values = localSettings->Values;
+    values->Remove(PlatformStringFromString(key));
```
Comment (zilongshanren) [inline cocos/base/CCUserDefault-winrt.cpp:342]: do we need call `flush` here?

File: cocos/base/CCUserDefault.h [hunk 5]
```diff
@@ -202,6 +202,14 @@ class CC_DLL UserDefault
      */
     virtual void flush();
 
+    /**
+    * delete any value by key, if the key doesn't exist, will return true.
```
Comment (zilongshanren) [inline cocos/base/CCUserDefault.h]: What does it mean "if the key doesn't exist, will return true"?
If the parameter is nullptr, you return false. If the key doesn't exist, you return false.
I think we can't tell these two cases apart, right?

File: cocos/platform/android/java/src/org/cocos2dx/lib/Cocos2dxHelper.java [hunk 6]
```diff
@@ -526,6 +526,13 @@ public static void setStringForKey(String key, String value) {
         editor.commit();
     }
     
+    public static void deleteValueForKey(String key) {
```
Comment (zilongshanren) [inline cocos/platform/android/java/src/org/cocos2dx/lib/Cocos2dxHelper.java:529]: This function should return bool? So we can tell whether it delete value successfully or not.

**Comment by jianglong0156** [comment]
arrange new pr:https://github.com/cocos2d/cocos2d-x/pull/12697

**Pull request closed without merging**